rustPlatform.fetchCargoVendor: allow duplicated dependencies

TomaSajt opened the pull request:
This is a breaking change (even if mostly internal). Because of this, it's not really backportable either.

Reviews are welcome.

Fixes https://github.com/NixOS/nixpkgs/issues/359340
Fixes https://github.com/NixOS/nixpkgs/issues/30742

Related:
- https://github.com/NixOS/nixpkgs/issues/30742
- https://github.com/NixOS/nixpkgs/issues/183344
- https://github.com/NixOS/nixpkgs/issues/359340
- https://github.com/NixOS/nixpkgs/pull/282798

---

This PR contains the simpler but arguably more important part of https://github.com/NixOS/nixpkgs/pull/282798.
The other PR tries to migrate both `importCargoLock` and `fetchCargoVendor`, but this one only does the `fetchCargoVendor` change.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

TomaSajt committed:
rustPlatform.fetchCargoVendor: allow duplicated dependencies

This is achieved by having subdirs in the vendor dir

TomaSajt committed:
treewide: fix search location of crates inside cargoDepsCopy

This is needed because the previous commit moved crates one
layer deeper in the vendor directory of rust packages when using
fetchCargoVendor

nyabinary commented:
Should target staging considering the amount of rebuilds

TomaSajt commented:
Yes I'll point it there soon, but I still have a few TODOs.

poelzi commented:
Unfortunately, this does not fix the sui problem. I debugged it by adding "set -x" on the cargo-vendor-dir command:

```
+++ ln -s /nix/store/a3hq2s4jsrjlk5flzm6bwibc63habljn-tokio-io-timeout-1.2.0 /nix/store/hrw1qsrqh3mfasw72hlapd0lnkd303a1-cargo-vendor-dir/tokio-io-timeout-1.2.0
+++ '[' -e /nix/store/a3hq2s4jsrjlk5flzm6bwibc63habljn-tokio-io-timeout-1.2.0/.cargo-config ']'
+++ for crate in /nix/store/w4njzhl6c2ap84kd6v9c09g3vb3jwpwr-Inflector-0.11.4 /nix/store/61myz4b2wg5bcwrmcgqir98p1kjh0yjf-addchain-0.2.0 /nix/store/dfdynm0vivlyw1cyxabpnkmk0gbfvn6v-add>
++++ basename /nix/store/pbkdlfdqi3gzigz428kbay8fcwvv6z7z-tokio-macros-2.5.0
++++ cut -c 34-
+++ ln -s /nix/store/pbkdlfdqi3gzigz428kbay8fcwvv6z7z-tokio-macros-2.5.0 /nix/store/hrw1qsrqh3mfasw72hlapd0lnkd303a1-cargo-vendor-dir/tokio-macros-2.5.0
+++ '[' -e /nix/store/pbkdlfdqi3gzigz428kbay8fcwvv6z7z-tokio-macros-2.5.0/.cargo-config ']'
+++ for crate in /nix/store/w4njzhl6c2ap84kd6v9c09g3vb3jwpwr-Inflector-0.11.4 /nix/store/61myz4b2wg5bcwrmcgqir98p1kjh0yjf-addchain-0.2.0 /nix/store/dfdynm0vivlyw1cyxabpnkmk0gbfvn6v-add>
++++ basename /nix/store/rk97wgy1a6m8x9z5k05dzw1fmhbqb3wx-tokio-macros-2.5.0
++++ cut -c 34-
+++ ln -s /nix/store/rk97wgy1a6m8x9z5k05dzw1fmhbqb3wx-tokio-macros-2.5.0 /nix/store/hrw1qsrqh3mfasw72hlapd0lnkd303a1-cargo-vendor-dir/tokio-macros-2.5.0
ln: failed to create symbolic link '/nix/store/hrw1qsrqh3mfasw72hlapd0lnkd303a1-cargo-vendor-dir/tokio-macros-2.5.0/rk97wgy1a6m8x9z5k05dzw1fmhbqb3wx-tokio-macros-2.5.0': Permission den>
+ exitHandler
+ exitCode=1
+ set +e
+ '[' -n '' ']'
+ ((  1 != 0  ))
+ runHook failureHook
+ local hookName=failureHook
```

The problem is triggered by 2 versions of the tokio-macros used by sui

TomaSajt commented:
You are not using `fetchCargoVendor` you are using `importCargoLock`.

I was able to build sui with this:
```nix
{
  lib,
  rustPlatform,
  fetchFromGitHub,
  pkg-config,
  protobuf,
  zstd,
}:

rustPlatform.buildRustPackage (finalAttrs: {
  pname = "sui";
  version = "1.43.1";

  src = fetchFromGitHub {
    owner = "MystenLabs";
    repo = "sui";
    tag = "mainnet-v${finalAttrs.version}";
    hash = "sha256-mNljh3HupusGmfT3pXtjqUp7OZHhyWd6jNiFiKlSYpk=";
  };

  useFetchCargoVendor = true;
  cargoHash = "sha256-ckotvpQw3WfvJ2YXR/XKT7LamGj7kLtGwMR/qrXpmYc=";

  nativeBuildInputs = [
    pkg-config
    #protobuf
    rustPlatform.bindgenHook
  ];

  buildInputs = [
    zstd
  ];

  env = {
    # if this is not set the build will try to invoke git to get the rev
    GIT_REVISION = "nixpkgs@${finalAttrs.version}";
    ZSTD_SYS_USE_PKG_CONFIG = true;
  };

  # the build takes up more than 60G of space even without the checks
  # my system ran out of space (150G) when having tests enabled
  doCheck = false;

})
```

poelzi commented:
Ahh, I see. I used `useFetchCargoVendor = true` and `cargoLock`.
Works now, thanks a lot 👍🏽

PerchunPak commented on pkgs/by-name/rq/rquickshare/package.nix:
```diff
@@ -50,7 +50,7 @@ rustPlatform.buildRustPackage rec {
+    substituteInPlace $cargoDepsCopy/*/libappindicator-sys-*/src/lib.rs \
       --replace-fail "libayatana-appindicator3.so.1" "${libayatana-appindicator}/lib/libayatana-appindicator3.so.1"
   '';
 
```
`cargoHash` if you remove the patch
```nix
  cargoHash = app-type-either "sha256-Blx64KyF0eTsNTqv9uDsXIbPyjhIb6POCZAYkjC9br0=" "sha256-oNbB2rkASGk8PO54t70ZukXcfJGrN5BxeMoQXZNCFVM=";
```

TomaSajt replied:
Thanks, removed the patch.

PerchunPak reviewed (approved):
`rquickshare` builds on aarch64-linux (without the patch), so LGTM

dotlambda commented:
> do NOT merge before we decide on a final design

Whose opinion would you like to hear before merging?

Since this is a breaking change, we have to wait until 2025-05-08 before merging according to https://github.com/NixOS/nixpkgs/issues/390768.

Btw, I successfully built [deltachat-tauri](https://github.com/deltachat/deltachat-desktop/tree/main/packages/target-tauri) using this branch: https://github.com/NixOS/nixpkgs/pull/398998

Bot-wxt1221 commented:
Stupid patch in rustdesk-flutter can be removed in this PR or in a later PR.

amozeo commented:
I found [cargo lock](https://raw.githubusercontent.com/pop-os/cosmic-applibrary/cec630b462e149f42a227fd38dee2d0b5658ab06/Cargo.lock) in the wild that has duplicated dependencies and changes in this pr don't fix it. I assume it's because duplicated depencency also uses same git revision.

<details><summary>Nix file</summary>
<p>

```nix
{
  nixpkgs ? builtins.fetchTarball {
    url = "https://github.com/NixOS/nixpkgs/archive/05c2c0fe38a362c3a27b45c85f0732acae693a6e.tar.gz";
    sha256 = "sha256-pAJAPMVlkxGUBsW0zO9mQFWXDmtCgS8M+KuhOh7rPAs=";
  },
  pkgs ? import (nixpkgs) {
    config = {};
    overlays = [];
  },
}:

let
  inherit (pkgs) rustPlatform fetchFromGitHub;
in

rustPlatform.fetchCargoVendor {
  src = fetchFromGitHub {
    owner = "pop-os";
    repo = "cosmic-applibrary";
    rev = "cec630b462e149f42a227fd38dee2d0b5658ab06";
    hash = "sha256-RXwf+JnP8ft+ELgPuPiNrWcKMnP2JNdhHw9tlm/CpoA=";
  };
  name = "vendor";
  hash = "sha256-NZh6KLukE31oAmCG0Qy5u8NPunjrHJ+3tUiw4kdI0y4=";
}
```

</p>
</details>

<details><summary>Error</summary>
<p>

```
Traceback (most recent call last):
  File "/nix/store/wmxh47lxmrga193iaacw1s6df5la0z44-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 327, in <module>
    main()
  File "/nix/store/wmxh47lxmrga193iaacw1s6df5la0z44-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 323, in main
    subcommand_func()
  File "/nix/store/wmxh47lxmrga193iaacw1s6df5la0z44-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 315, in <lambda>
    "create-vendor": lambda: create_vendor(vendor_staging_dir=Path(sys.argv[2]), out_dir=Path(sys.argv[3]))
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/wmxh47lxmrga193iaacw1s6df5la0z44-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 265, in create_vendor
    copy_and_patch_git_crate_subtree(git_tree, pkg["name"], crate_out_dir)
  File "/nix/store/wmxh47lxmrga193iaacw1s6df5la0z44-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 202, in copy_and_patch_git_crate_subtree
    shutil.copytree(crate_tree, crate_out_dir, ignore_dangling_symlinks=True)
  File "/nix/store/f2krmq3iv5nibcvn4rw7nrnrciqprdkh-python3-3.12.9/lib/python3.12/shutil.py", line 600, in copytree
    return _copytree(entries=entries, src=src, dst=dst, symlinks=symlinks,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/f2krmq3iv5nibcvn4rw7nrnrciqprdkh-python3-3.12.9/lib/python3.12/shutil.py", line 498, in _copytree
    os.makedirs(dst, exist_ok=dirs_exist_ok)
  File "<frozen os>", line 225, in makedirs
FileExistsError: [Errno 17] File exists: '/nix/store/xm8g8zkzbnaa83wx698z58g053vya21n-vendor-vendor/git-e838616813470aa6a108ede07ab8e6050b68a5a6/cosmic-config-0.1.0'
```

</p>
</details>

TomaSajt commented:
Nice catch.
It shouldn't be too hard to fix this.

amozeo commented:
quickly looking into how to resolve this I think it would be good idea to enumerate `cargo_lock_toml["package"]` in [line 224](https://github.com/NixOS/nixpkgs/pull/387337/files#diff-77d6ed0c1e97de400495829ad6c81d6da127ff83ff09d6d14ec9f887f46d4c67R244) and use index to generate subdirectories names

TomaSajt commented:
The exact same package is present in the lockfile two times, but via a slightly different `source` value.

`git+https://github.com/pop-os/libcosmic//?rev=e838616#e838616813470aa6a108ede07ab8e6050b68a5a6`
`git+https://github.com/pop-os/libcosmic/#e838616813470aa6a108ede07ab8e6050b68a5a6`

We could:
 1. have special handling for git deps:
 don't do the copying step if the `(pname, version, git_rev_sha`) tuple has already been copied.
 2. actually make the two different sources use different directories
 (maybe via enumeration: `source-git-0`, `source-git-1`, `source-git-2`)

> quickly looking into how to resolve this I think it would be good idea to enumerate `cargo_lock_toml["package"]` in [line 224](https://github.com/NixOS/nixpkgs/pull/387337/files#diff-77d6ed0c1e97de400495829ad6c81d6da127ff83ff09d6d14ec9f887f46d4c67R244) and use index to generate subdirectories names

AFAIK we only have the luxury of overriding entire sources and not packages themselves.
It would be pretty cool if we could have one directory for each package, (probably via enumeration).
(We *might* be able to do this by editing the lockfile but I do not want to thread those waters if possible)

---

IMO the second option seems like a more general solution, that could eventually also support multiple registries via: `source-registry-0` `source-registry-1`

(We could also just not have the directory names contain source-type: `source-0`, `source-1` etc.)

TomaSajt commented:
Pushed a commit which makes it so that each `source` has its own directory in `$out/source-$ind` where `$ind` is assigned to a source based on order of appearance.
Because `source.crates-io` is already internally defined, it deserves special handling.
To facilitate this, I reserved it to be `ind = 0`.

Reminder: this does not touch the FOD part of the package: meaning that we don't have to worry about breaking any hashes.

TomaSajt commented:
Example: the layout of `air-formatter.cargoDeps`:
```
$ ls -A result/
.cargo  source-0  source-1  source-2  source-3  Cargo.lock
```

```
$ cat result/.cargo/config.toml

[source.nixpkgs-vendored-source-0]
directory = "@vendor@/source-0"

[source.crates-io]
replace-with = "nixpkgs-vendored-source-0"

[source.nixpkgs-vendored-source-1]
directory = "@vendor@/source-1"

[source.nixpkgs-original-source-1]
replace-with = "nixpkgs-vendored-source-1"
git = "https://github.com/biomejs/biome"
rev = "2648fa4201be4afd26f44eca1a4e77aac0a67272"

[source.nixpkgs-vendored-source-2]
directory = "@vendor@/source-2"

[source.nixpkgs-original-source-2]
replace-with = "nixpkgs-vendored-source-2"
git = "https://github.com/lionel-/tower-lsp"
branch = "bugfix/patches"

[source.nixpkgs-vendored-source-3]
directory = "@vendor@/source-3"

[source.nixpkgs-original-source-3]
replace-with = "nixpkgs-vendored-source-3"
git = "https://github.com/r-lib/tree-sitter-r"
rev = "a0d3e3307489c3ca54da8c7b5b4e0c5f5fd6953a"

```

nyabinary commented:
The Zed update is reliant on this change iirc unfortunately

TomaSajt commented:
New format:
```
$ ls -A result/
.cargo  source-git-0  source-git-1  source-git-2  source-registry-0  Cargo.lock
```

```
$ cat result/.cargo/config.toml

[source._vendored-source-registry-0]
directory = "@vendor@/source-registry-0"

[source.crates-io]
replace-with = "_vendored-source-registry-0"

[source._vendored-source-git-0]
directory = "@vendor@/source-git-0"

[source._original-source-git-0]
git = "https://github.com/biomejs/biome"
rev = "2648fa4201be4afd26f44eca1a4e77aac0a67272"
replace-with = "_vendored-source-git-0"

[source._vendored-source-git-1]
directory = "@vendor@/source-git-1"

[source._original-source-git-1]
git = "https://github.com/lionel-/tower-lsp"
branch = "bugfix/patches"
replace-with = "_vendored-source-git-1"

[source._vendored-source-git-2]
directory = "@vendor@/source-git-2"

[source._original-source-git-2]
git = "https://github.com/r-lib/tree-sitter-r"
rev = "a0d3e3307489c3ca54da8c7b5b4e0c5f5fd6953a"
replace-with = "_vendored-source-git-2"
```

DavHau reviewed (commented):
I tested this patch today on a project where it was running into the `failed to create symbolic link` error and it fixed it successfully.

EDIT: Though I'm now running into an error like this:
```
 error: failed to select a version for the requirement `cfg-if = "^1.0"` (locked to 1.0.0)
```

Not sure if this is related to the fetching or not

EDIT: forget about it. This seems to be an unrelated issue

MultisampledNight commented:
Packaging [Kiesel](https://kiesel.dev) seems to be blocked on this.

Kiesel uses an in-tree Rust crate for some FFI, which depends partly on crates.io versions, partly on git versions to get the latest fixes. A lot of transitive deps are duplicated in turn, too. 
Just replacing them with crates.io is not practical, as the rest of the project also has the exact git commits pinned.
(Current attempt is at https://github.com/MultisampledNight/nixpkgs/commit/ec0f801a729aee5df2fbf4f6eea49803b525aca3 if anybody is curious, very half-baked.)

MultisampledNight reviewed (approved):
Simple enough impl, impact is bearable.

jeafleohj committed:
rustPlatform.fetchCargoVendor: make manifest discovery deterministic

Sort Cargo.toml paths by depth/name for traversal.
Skip manifests where cargo metadata fails to avoid aborting on
invalid or unsupported example crates.
Fix mainfest/manifest typo in helper name.
